### ISSUE TITLE
add dependsOn functionallity

### ISSUE THREAD

**Issue by timomayer** [virtualidentityag/conditional-resource-loader#7]
the new dependsOn array defines dependencies. if it is set the scripts should only start to load when all scripts in the dependsOn array are already available.

make sure if a file in dependsOn array does not exist in loading queue that the dependency is skipped and the files are loaded

**Comment by timomayer** [comment]
#9

**Comment by smohadjer** [comment]
@Timo: 
1. Do we need dependency for .css files, or only for .js files?
2. Do we need to support dependencies more than one level, that is can dependencies themselves have dependencies?

**Comment by smohadjer** [comment]
I find the data-resources syntax that we are using for handling dependencies bad as when we have multiple resources it's not clear to which resource the dependency belongs and this not only causes delay in loading of some resources, it can also break our loader script in some cases. For example suppose component A needs a.js and b.js and component B needs c.js and d.js. And let's suppose c.js is a dependency for a.js and b.js is a dependency for d.js. Now due to our ambiguous data-resources syntax, our loader script will think that c.js is a dependency for b.js (from component A) and that b.js is a dependency for c.js (from component B), causing a situation where it's impossible to determine whether c.js should load first or b.js.

My suggestion is that we change syntax for dependency array so that there would be no ambiguity, perhaps using a syntax like this: 

dependency: [
    {script: [dependency1, dependency2]}
]

**Comment by alxbenz** [comment]
Hi @smohadjer,

I tried to clarify your example by providing the current syntax for the two components:

Component A 
`[
  {
      paths: ['a.js', 'c.js']
  },
  {
      paths: ['b.js'],
      dependsOn: ['c.js']
  }
]`

Component B 
`[
  {
      paths: ['c.js', 'b.js']
  },
  {
      paths: ['d.js'],
      dependsOn: ['b.js']
  }
]`

You are right, that this would be a situation where it is kind of ambiguous what happens. 
But is this even a valid example? Why would script b.js depend on c.js for one component but not for the other. So the correct Syntax for Component B should be:

Component B 
`[
  {
      paths: ['c.js']
  },
  {
      paths: ['d.js'],
      dependsOn: ['b.js']
  }, 
  {
	paths: ['b.js'],
	dependsOn: ['c.js']
  }
]`

Making the load order: a.js, c.js -> b.js -> d.js
Could you provide an example with your suggested syntax?

Concerning your other questions:
1. I don't think we need a dependency for css files.
2. Yes they can and it is already implemented like that.

**Comment by smohadjer** [comment]
Can't my example be written like this?

Component A (a.js depends on c.js)
`[
  {
      paths: ['a.js', 'b.js'],
      dependsOn: ['c.js']
  }
]`

Component B (d.js depends on b.js)
`[
  {
      paths: ['c.js', 'd.js'],
      dependsOn: ['b.js']
  }
]`

**Comment by alxbenz** [comment]
By writing it like that it would mean:
- for Component A: a.js and b.js depend on c.js
- for Component B: c.js and b.js depend in b.js

This would result in a cyclic dependency and you would get a warning. None of the scripts could be loaded.
Also you should put the dependencies in an extra path, since they are not loaded automatically.

**Comment by smohadjer** [comment]
I didn't know that dependencies would apply to "ALL" resources listed in the paths array, specially since in demo page you have an example like this: 

`{
paths: ['js/jquery.component.js', 'js/jquery.component3.js', '/demo/resources-content/css/demo4.css'],
dependsOn: ['js/jquery.component2.js', '/demo/resources/js/jquery.plugin.js']
}
`

Are "jquery.component2.js" and "jquery.plugin.js" in example above dependency for all the resources listed in the paths array?

**Comment by smohadjer** [comment]
I also don't understand your comment below?

> Also you should put the dependencies in an extra path, since they are not loaded automatically.

 Are you saying if I have just one resource which has only one dependency, I have to write that in two different paths object? In other words is below syntax wrong?

`
			 data-resources="
			 [
			 	{
			 		paths: ['js/jquery.component2.js'],
			 		dependsOn: ['js/jquery.component1.js']
			 	}
			 ]">
`

**Comment by alxbenz** [comment]
Yes dependsOn is for all resources listed in the according resources array.

´data-resources="[ { paths: ['js/jquery.component2.js'], dependsOn: ['js/jquery.component1.js'] } ´

This is wrong, because the js/jquery.component1.js is not loading on its own.

I will clean up the demo.html to clarify the syntax.

**Comment by smohadjer** [comment]
So instead of: 
`
data-resources=" [ { paths: ['js/jquery.component2.js'], dependsOn: ['js/jquery.component1.js'] } ]">
`

We have to write:
`
data-resources=" [ { paths: ['js/jquery.component1.js']}, { paths: ['js/jquery.component2.js'], dependsOn: ['js/jquery.component1.js'] } ]">
`

I don't find this syntax intuitive or even clear, so I suggest if you want to go with it at least write the rules for how dependencies should be written in README file. It seems even Timo understood the syntax wrong as the way he updated the syntax in README file now there is a cyclic dependency for component2.js (in 2nd path component.js is dependency for component2.js, but in 3rd path the opposite is the case)!

**Comment by timomayer** [comment]
@smohadjer @alxbenz described exactly how i planned it. dependsOn is only a defintion for dependencies, files in there should not load automatically. if that would be the case there is no possibility to define dependency for jquery.component1.js because it would be loaded because it is defined as a dependency for js/jquery.component2.js

**Comment by smohadjer** [comment]
Not necessarily true, that depends on how the loader logic is written. The loader can parse all data-resources on a page and create a complete dependency tree before starting to load files. This way dependencies can have their own dependencies.

**Comment by timomayer** [comment]
yes but is think that makes it even more unintuitive. we should add a part to the documentation to make it clear that dependsOn on defines dependencies and does not automatically load the files in there.

**Comment by smohadjer** [comment]
Not sure what you mean by depends on does not automatically load the files, that's not what I suggested. In my opinion if we have a component in which one script depends on another script, we need only one resource object, not two, same as I wrote in my last comment to Alex above. The part I don't understand is why one resource is not enough.

Nevertheless, the way you updated syntax usage in documentation is wrong, since according to Alex dependency scripts apply to ALL scripts in corresponding path array. In README.md now we have a dependency loop between component1.js and component2.js. https://github.com/virtualidentityag/conditional-resource-loader/pull/9/commits/9143fb6100dff680e5f9aa1afff31398a985d5ff
